src/main.c

```diff
@@ -130,7 +130,162 @@ static void import() {
 	fclose(f);
 }
 
-/*static void import_spc() {
+struct spcDetails {
+	WORD music_table_addr;
+	WORD instrument_table_addr;
+	WORD music_addr;
+	BYTE music_index;
+};
+
+BOOL try_parse_music_table(const BYTE *spc, struct spcDetails *out_details) {
+	if (memcmp(spc, "\x1C\x5D\xF5", 3) != 0) return FALSE;
+	WORD addr_hi = *((WORD*)&spc[3]);
+
+	// Check for Konami-only branch
+	if (spc[5] == 0xF0) {
+		spc += 2; // skip two bytes (beq ..)
+	}
+
+	if (spc[5] != 0xFD) return FALSE;
+
+	// Check for Starfox branch
+	if (memcmp(&spc[6], "\xD0\x03\xC4", 3) == 0 && spc[10] == 0x6F) {
+		spc += 5; // skip these 5 bytes
+	}
+
+	if (spc[6] != 0xF5) return FALSE; // mov a,
+	WORD addr_lo = *((WORD*)&spc[7]); //       $....+x
+
+	if (spc[9] != 0xDA || spc[10] != 0x40) return FALSE; // mov $40,ya
+
+	// Validate retrieved address
+	if (addr_lo != addr_hi - 1) return FALSE;
+
+	out_details->music_table_addr = addr_lo;
+	return TRUE;
+}
+
+BOOL try_parse_music_address(const BYTE* spc, struct spcDetails *out_details) {
+	WORD loop_addr = *(WORD *)&spc[0x40];
+	WORD *terminator = (WORD *)&spc[loop_addr];
+	while (terminator[0]) terminator++;
+
+	// sanity check (abitrarily limit pattern count to 256)
+	if (terminator - (WORD *)&spc[loop_addr] > 0x200) return FALSE;
```

Comment:
The code as it stands doesn't do a great job of enforcing the 256 pattern limit; it can read out of bounds, which might crash ebmused.

```suggestion
	const WORD *terminator;
	{
		int i = 0;
		while (i < 256 // sanity check (arbitrarily limit pattern count to 256)
				&& loop_addr + 2*i + 1 <= 0x10000 - 16 - 2 // We need at least 16 bytes after the terminator for a pattern, and the terminator itself is two bytes
				&& spc[loop_addr + 2*i] != 0 && spc[loop_addr + 2*i + 1] != 0) {
			i++;
		}
		terminator = (const WORD *)&spc[loop_addr + 2*i];

		if (i == 256 || loop_addr + 2*i > 0x10000 - 16 - 2 || *terminator != 0) return FALSE;
	}
```

(...There might be a better way to write that that doesn't involve creating a new scope or that has more/fewer pointer casts, but eh)

Checking for out of bounds reads is complicated. The loop below would also have to be modified to completely fix it. Perhaps there's a better solution with fewer moving parts...

Reply:
This logic's been updated to check/prevent out-of-bounds access